src/coord/src/coord.rs

```diff
@@ -252,6 +248,56 @@ pub struct Coordinator {
     ///
     /// The responses have the form `Vec<Row>` but should perhaps become `TailResponse`.
     pending_tails: HashMap<GlobalId, mpsc::UnboundedSender<Vec<Row>>>,
+
+    /// Tracks timestamps per timeline to provide linearizability
+    /// guarantees. SQL statements that determine timestamps should call
+    /// `advance_or_return_timestamp` to ensure timestamps never decrease across
+    /// transactions.
+    timelines: HashMap<TimelineId, Timeline>,
+}
+
+/// A Timeline provides linearizability to callers by enforcing relationships
+/// between reads and writes.
+struct Timeline {
+    /// The timeline's current timestamp.
+    ts: Timestamp,
+    /// Whether the most recent operation was a read.
+    last_op_was_read: bool,
+}
+
+impl Timeline {
+    fn new(ts: Timestamp) -> Self {
+        Self {
+            ts,
+            last_op_was_read: false,
+        }
+    }
+
+    /// Assign a timestamp for a read.
+    fn get_read_ts(&mut self) -> Timestamp {
+        self.last_op_was_read = true;
+        self.ts
+    }
+
+    /// Assign a timestamp for a write. Writes following reads must ensure that
+    /// they are assigned a strictly larger timestamp to ensure they are not
+    /// visible to any real-time earlier reads.
+    fn get_write_ts(&mut self) -> Timestamp {
+        if self.last_op_was_read {
+            self.last_op_was_read = false;
+            self.ts += 1;
+        }
+        self.ts
+    }
+
+    /// Advance `ts` to `now` if `now > ts`. Use when the Timeline may have
+    /// advanced and we need to force observation of new timestamps.
+    fn maybe_advance(&mut self, now: Timestamp) {
```

Comment:
Yeah, I think now with some of the framing above, this can be more clearly (to me, at least) explained as "advances to writes at `now` if it has not yet occurred; if reads at `now` have already occurred, or the timeline timestamp already passed `now`, this call has no effect". Something sort like that, I suppose!